native/ios/WatermelonDB/Database.cpp

```diff
@@ -1 +1,533 @@
+    } catch (const std::exception &ex) {
+        rollback();
+        throw;
+    }
+
+    for (auto const &key : addedIds) {
+        markAsCached(key);
+    }
+
+    for (auto const &key : removedIds) {
+        removeFromCache(key);
+    }
+}
+
+jsi::Array Database::getDeletedRecords(jsi::String &tableName) {
+    auto &rt = getRt();
+    auto args = jsi::Array::createWithElements(rt);
+    auto statement = executeQuery("select id from " + tableName.utf8(rt) + " where _status='deleted'", args);
+
+    jsi::Array records(rt, 0);
+
+    for (size_t i = 0; true; i++) {
+        int stepResult = sqlite3_step(statement.stmt);
+
+        if (stepResult == SQLITE_DONE) {
+            break;
+        } else if (stepResult != SQLITE_ROW) {
+            throw dbError("Failed to get deleted records");
+        }
+
+        assert(sqlite3_data_count(statement.stmt) == 1);
+
+        const char *idText = (const char *)sqlite3_column_text(statement.stmt, 0);
+        if (!idText) {
+            throw jsi::JSError(rt, "Failed to get ID of a record");
+        }
+
+        jsi::String id = jsi::String::createFromAscii(rt, idText);
+        records.setValueAtIndex(rt, i, id);
+    }
+
+    return records;
+}
+
+void Database::destroyDeletedRecords(jsi::String &tableName, jsi::Array &recordIds) {
+    auto &rt = getRt();
+    beginTransaction();
+    try {
+        // TODO: Maybe it's faster & easier to do it in one query?
+        std::string sql = "delete from " + tableName.utf8(rt) + " where id == ?";
+
+        for (size_t i = 0, len = recordIds.size(rt); i < len; i++) {
+            // TODO: What's the behavior if record doesn't exist or isn't actually deleted?
+            jsi::String id = recordIds.getValueAtIndex(rt, i).getString(rt);
+            auto args = jsi::Array::createWithElements(rt, id);
+            executeUpdate(sql, args);
+        }
+        commit();
+    } catch (const std::exception &ex) {
+        rollback();
+        throw;
+    }
+}
+
+const std::string localStorageSchema = R"(
+create table local_storage (
+key varchar(16) primary key not null,
+value text not null
+);
+
+create index local_storage_key_index on local_storage (key);
+)";
+
+void Database::unsafeResetDatabase(jsi::String &schema, int schemaVersion) {
+    auto &rt = getRt();
+    beginTransaction();
+    try {
+        // TODO: delete file in non-test?
+
+        std::vector<std::string> tables = {};
+
+        // Find all tables (scope to reset SqliteStatement)
+        {
+            auto args = jsi::Array::createWithElements(rt);
+            auto statement = executeQuery("select name from sqlite_master where type='table'", args);
+
+            for (size_t i = 0; true; i++) {
+                int stepResult = sqlite3_step(statement.stmt);
+
+                if (stepResult == SQLITE_DONE) {
+                    break;
+                } else if (stepResult != SQLITE_ROW) {
+                    throw dbError("Failed to get table names to delete");
```

Comment:
This for loop with checking stepResult happens a few times. Maybe move to a helper method?

Reply:
yeah I was thinking about it, but making a helper just for _this_ case doesn't seem like worth the trouble, and there's a few ways that steps are treated - sometimes we allow DONE, and break or treat it as null, and sometimes it's an error. So I'm not sure how to best abstract it...

Reply:
Maybe you're right. If it have to use 4-5 args then maybe it's not worth it :)